x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java

```diff
@@ -403,21 +411,56 @@ private void startComputeOnRemoteClusters(
                     queryPragmas.exchangeBufferSize(),
                     esqlExecutor,
                     refs.acquire().delegateFailureAndWrap((l, unused) -> {
-                        var remoteSink = exchangeService.newRemoteSink(rootTask, childSessionId, transportService, cluster.connection);
-                        exchangeSource.addRemoteSink(remoteSink, true, queryPragmas.concurrentExchangeClients(), ActionListener.noop());
                         var remotePlan = new RemoteClusterPlan(plan, cluster.concreteIndices, cluster.originalIndices);
                         var clusterRequest = new ClusterComputeRequest(cluster.clusterAlias, childSessionId, configuration, remotePlan);
-                        var clusterListener = ActionListener.runBefore(
-                            computeListener.acquireCompute(cluster.clusterAlias()),
-                            () -> l.onResponse(null)
+                        // Create group task for this cluster. This group task ensures that two branches of the computation:
+                        // the exchange sink and the cluster request, belong to the same group and each of them can cancel the other.
+                        // runAfter listeners below ensure that the group is finalized when both branches are done.
+                        // The group task is the child of the root task, so if the root task is cancelled, the group task is cancelled too.
+                        CancellableTask groupTask = createGroupTask(rootTask, clusterRequest.getDescription());
```

Comment:
Can we encapsulate this into ComputerListener or somewhere else to make this path more isolated and easier to follow?

Reply:
I moved the group task code into a separate class for now.